src/_pytest/main.py

```diff
@@ -523,6 +503,43 @@ def gethookproxy(self, fspath: py.path.local):
             proxy = self.config.hook
         return proxy
 
+    def _recurse(self, direntry: "os.DirEntry[str]") -> bool:
+        if direntry.name == "__pycache__":
+            return False
+        path = py.path.local(direntry.path)
+        ihook = self.gethookproxy(path.dirpath())
+        if ihook.pytest_ignore_collect(path=path, config=self.config):
+            return False
+        norecursepatterns = self.config.getini("norecursedirs")
+        for pat in norecursepatterns:
+            if path.check(fnmatch=pat):
+                return False
+        return True
```

Comment:
Thanks! Applied this and the if-combining suggestion.

Reply:
I think you can just return negated `any()` instead of having it in the `if` clause:  
```python
return not any(path.check(fnmatch=pat) for pat in norecursepatterns)
```